var/spack/repos/builtin/packages/vvtest/package.py

```diff
@@ -0,0 +1,30 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class Vvtest(Package):
+    """Vvtest is a test harness originally authored by Richard Drake for the
+    Alegra-Emphasis project, written in pure Python, and designed for running
+    verification, validation, and regression type tests. Being a stand-alone
+    product, it can also be used by analysts to help manage sets of
+    simulations, especially useful on batch platforms."""
+
+    homepage    = "https://github.com/rrdrake/vvtest"
+    url         = "https://github.com/rrdrake/vvtest/archive/1.0.0.tar.gz"
+    git         = "https://github.com/rrdrake/vvtest.git"
+    maintainers = ['mrmundt', 'rrdrake']
+
+    version('1.2.0', sha256='d6b2432a2e6c43fb0d87ffc0eaa34a74d2268a732f7709ebdcf1344fbcaee154')
+    version('1.1.0', sha256='674585f12d393ab9745a5ab26f59cb0f0e213f9c597b37467125979b5955ca79')
+    version('1.0.0', sha256='acd04e8e6635ed1b1725b793e8287a58831d6380759a81159142a6ff3397a8dd')
+
+    depends_on('python@2.6.0:3')
+
+    def install(self, spec, prefix):
+        set_executable('install_vvtest')
+        python = which('python')
```

Comment:
Python is confusing, since there might not be a `python` executable, only `python2` or `python3`. In this case, let's use:
```suggestion
        python = spec['python'].command
```

Reply:
With this one, I end up getting an error:
```
KeyError: 'No spec with name python in vvtest@1.2.0%apple-clang@11.0.3 arch=darwin-catalina-x86_64'
```
Clearly, this is on a Mac. But it works just fine on a RHEL. Any ideas as to why Mac might be weird?

Reply:
That error means you're missing a dependency on `python`. Are you on the latest commit of your PR on macOS?

Reply:
Yup. I've got the same copy on my Mac and my RHEL:
```
from spack import *

class Vvtest(Package):
    """Vvtest is a test harness originally authored by Richard Drake for the
    Alegra-Emphasis project, written in pure Python, and designed for running
    verification, validation, and regression type tests. Being a stand-alone
    product, it can also be used by analysts to help manage sets of
    simulations, especially useful on batch platforms."""

    homepage    = "https://github.com/rrdrake/vvtest"
    url         = "https://github.com/rrdrake/vvtest/archive/1.0.0.tar.gz"
    git         = "https://github.com/rrdrake/vvtest.git"
    maintainers = ['mrmundt', 'rrdrake']

    version('1.2.0', sha256='d6b2432a2e6c43fb0d87ffc0eaa34a74d2268a732f7709ebdcf1344fbcaee154')
    version('1.1.0', sha256='674585f12d393ab9745a5ab26f59cb0f0e213f9c597b37467125979b5955ca79')
    version('1.0.0', sha256='acd04e8e6635ed1b1725b793e8287a58831d6380759a81159142a6ff3397a8dd')

    depends_on('python@2.6.0:3', type=('build', 'run'))

    def install(self, spec, prefix):
        python = spec['python'].command
        python('install_vvtest', prefix.bin)
```
Do you think it might be better to just try running `exec(open('./install_vvtest').read())`?

Reply:
Hmm, that doesn't make any sense. Can you run with `spack --debug` and see where this error is coming from?

Reply:
```
$ spack --debug install
==> [2020-09-17-10:48:55.786204] Imported install from built-in commands
==> [2020-09-17-10:48:55.787840] Reading config file /Users/mmundt/tpl_building/sems-spack-tpls/src/platform/dev/packages.yaml
==> [2020-09-17-10:48:55.790250] Reading config file /Users/mmundt/tpl_building/sems-spack-tpls/src/platform/dev/config.yaml
==> [2020-09-17-10:48:55.794679] Reading config file /Users/mmundt/tpl_building/sems-spack-tpls/src/project/dev/repos.yaml
==> [2020-09-17-10:48:55.797403] Reading config file /Users/mmundt/tpl_building/sems-spack-tpls/src/core/modules.yaml
==> [2020-09-17-10:48:55.815925] Reading config file /Users/mmundt/tpl_building/sems-spack-tpls/src/project/dev/spack.yaml
==> [2020-09-17-10:48:55.839961] Imported install from built-in commands
==> [2020-09-17-10:48:55.898899] Warning: included configuration files should be updated manually [files=../../core/modules.yaml, repos.yaml, ../../platform/${SEMS_MACHINE_NAME}/config.yaml, ../../platform/${SEMS_MACHINE_NAME}/packages.yaml]
==> [2020-09-17-10:48:55.932318] Installing environment /Users/mmundt/tpl_building/sems-spack-tpls/src/project/dev
==> [2020-09-17-10:48:55.934271] DATABASE LOCK TIMEOUT: 3s
==> [2020-09-17-10:48:55.934346] PACKAGE LOCK TIMEOUT: No timeout
==> [2020-09-17-10:48:55.935268] Reading config file /projects/sems/spack/etc/spack/defaults/repos.yaml
==> [2020-09-17-10:48:56.081465] Initializing the build queue for vvtest
==> [2020-09-17-10:48:56.081940] Bootstrapping apple-clang@11.0.3 compiler for vvtest
==> [2020-09-17-10:48:56.082238] Reading config file /Users/mmundt/.spack/darwin/compilers.yaml
==> [2020-09-17-10:48:56.107509] Removing failure marking for vvtest
==> [2020-09-17-10:48:56.108862] Acquiring a write lock on vvtest with timeout 1e-09
==> [2020-09-17-10:48:56.111602] Creating stage lock spack-stage-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2
==> [2020-09-17-10:48:56.152491] Installing vvtest
==> [2020-09-17-10:48:56.152574] Searching for binary cache of vvtest
==> [2020-09-17-10:48:56.152852] Reading config file /projects/sems/spack/etc/spack/defaults/mirrors.yaml
==> [2020-09-17-10:48:56.154835] Finding buildcaches at https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/build_cache
==> [2020-09-17-10:48:56.155151] Creating stage lock build_cache
==> [2020-09-17-10:48:56.155930] Checking existence of https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/build_cache/darwin-catalina-x86_64-apple-clang-11.0.3-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2.spec.yaml
==> [2020-09-17-10:48:56.156512] '/usr/bin/curl' '--stderr' '-' '-s' '-f' '-r' '0-0' 'https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/build_cache/darwin-catalina-x86_64-apple-clang-11.0.3-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2.spec.yaml'
==> [2020-09-17-10:48:56.388048] FailedDownloadError: Failed to fetch file from URL: https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/build_cache/darwin-catalina-x86_64-apple-clang-11.0.3-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2.spec.yaml
==> [2020-09-17-10:48:56.388241] Fetching from https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/build_cache/darwin-catalina-x86_64-apple-clang-11.0.3-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2.spec.yaml failed.
==> [2020-09-17-10:48:56.389255] No binary for vvtest found: installing from source
==> [2020-09-17-10:48:56.390262] Reading config file /projects/sems/spack/etc/spack/defaults/packages.yaml
==> [2020-09-17-10:48:56.413579] Reading config file /projects/sems/spack/etc/spack/defaults/darwin/packages.yaml
==> [2020-09-17-10:48:56.575396] [TOOLCHAIN] MATCH AppleClang
==> [2020-09-17-10:48:56.576152] [TOOLCHAIN] MATCH Clang
==> [2020-09-17-10:48:56.576718] [TOOLCHAIN] MATCH Gcc
==> [2020-09-17-10:48:56.577285] [TOOLCHAINS] {'Clang', 'Gcc', 'AppleClang'}
==> [2020-09-17-10:48:56.577915] Warning: microarchitecture specific optimizations are not supported yet on mixed compiler toolchains [check apple-clang@11.0.3 for further details]
==> [2020-09-17-10:48:56.621267] A dependency has updated CPATH, this may lead pkg-config to assume that the package is part of the system includes and omit it when invoked with '--cflags'.
==> [2020-09-17-10:48:56.724873] '/usr/bin/clang++' '-v' '/var/folders/zz/zyxvpxvq6csfxvn_n000_dw8002lg2/T/spack-implicit-link-infotbh5kbx1/main.c' '-o' '/var/folders/zz/zyxvpxvq6csfxvn_n000_dw8002lg2/T/spack-implicit-link-infotbh5kbx1/output'
==> [2020-09-17-10:48:56.787388] parsing implicit link info
==> [2020-09-17-10:48:56.788294] linker line:  "/Library/Developer/CommandLineTools/usr/bin/ld" -demangle -lto_library /Library/Developer/CommandLineTools/usr/lib/libLTO.dylib -no_deduplicate -dynamic -arch x86_64 -platform_version macos 10.15.0 10.15.4 -syslibroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -o /var/folders/zz/zyxvpxvq6csfxvn_n000_dw8002lg2/T/spack-implicit-link-infotbh5kbx1/output /var/folders/zz/zyxvpxvq6csfxvn_n000_dw8002lg2/T/main-387761.o -L/usr/local/lib -lc++ -lSystem /Library/Developer/CommandLineTools/usr/lib/clang/11.0.3/lib/darwin/libclang_rt.osx.a
==> [2020-09-17-10:48:56.788858] linkdir: /usr/local/lib
==> [2020-09-17-10:48:56.789387] found raw link dirs: /usr/local/lib
==> [2020-09-17-10:48:56.789982] found link dirs: /usr/local/lib
==> [2020-09-17-10:48:56.797422] Using cached archive: /projects/sems/spack/var/spack/cache/_source-cache/archive/d6/d6b2432a2e6c43fb0d87ffc0eaa34a74d2268a732f7709ebdcf1344fbcaee154.tar.gz
==> [2020-09-17-10:48:56.801913] Staging archive: /var/folders/zz/zyxvpxvq6csfxvn_n000_dw8002lg2/T/mmundt/spack-stage/spack-stage-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2/1.2.0.tar.gz
==> [2020-09-17-10:48:56.803593] '/usr/bin/tar' '-oxf' '/var/folders/zz/zyxvpxvq6csfxvn_n000_dw8002lg2/T/mmundt/spack-stage/spack-stage-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2/1.2.0.tar.gz'
==> [2020-09-17-10:48:56.926810] Created stage in /var/folders/zz/zyxvpxvq6csfxvn_n000_dw8002lg2/T/mmundt/spack-stage/spack-stage-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2
==> [2020-09-17-10:48:56.927964] No patches needed for vvtest
==> [2020-09-17-10:48:56.928515] vvtest: Building vvtest [Package]
==> [2020-09-17-10:48:56.936357] vvtest: Executing phase: 'install'
==> [2020-09-17-10:48:56.952072] Error: KeyError: 'No spec with name python in vvtest@1.2.0%apple-clang@11.0.3 arch=darwin-catalina-x86_64'

/Users/mmundt/tpl_building/sems-spack-tpls/src/project/dev/dev-repo/packages/vvtest/package.py:27, in install:
         26    def install(self, spec, prefix):
  >>     27        python = spec['python'].command
         28        python('install_vvtest', prefix.bin)

See build log for details:
  /var/folders/zz/zyxvpxvq6csfxvn_n000_dw8002lg2/T/mmundt/spack-stage/spack-stage-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2/spack-build-out.txt

Traceback (most recent call last):
  File "/projects/sems/spack/lib/spack/spack/spec.py", line 3258, in __getitem__
    value = next(
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/projects/sems/spack/lib/spack/spack/build_environment.py", line 847, in child_process
    return_value = function()
  File "/projects/sems/spack/lib/spack/spack/installer.py", line 1149, in build_process
    phase(pkg.spec, pkg.prefix)
  File "/projects/sems/spack/lib/spack/spack/package.py", line 109, in phase_wrapper
    phase(spec, prefix)
  File "/Users/mmundt/tpl_building/sems-spack-tpls/src/project/dev/dev-repo/packages/vvtest/package.py", line 28, in install
    python('install_vvtest', prefix.bin)
  File "/projects/sems/spack/lib/spack/spack/spec.py", line 3267, in __getitem__
    raise KeyError("No spec with name %s in %s" % (name, self))
KeyError: 'No spec with name python in vvtest@1.2.0%apple-clang@11.0.3 arch=darwin-catalina-x86_64'
==> [2020-09-17-10:48:56.952775] Flagging vvtest as failed: KeyError: 'No spec with name python in vvtest@1.2.0%apple-clang@11.0.3 arch=darwin-catalina-x86_64'
/Users/mmundt/tpl_building/sems-spack-tpls/src/project/dev/dev-repo/packages/vvtest/package.py:27, in install:
         26    def install(self, spec, prefix):
  >>     27        python = spec['python'].command
         28        python('install_vvtest', prefix.bin)

See build log for details:
  /var/folders/zz/zyxvpxvq6csfxvn_n000_dw8002lg2/T/mmundt/spack-stage/spack-stage-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2/spack-build-out.txt
Traceback (most recent call last):
  File "/projects/sems/spack/lib/spack/spack/spec.py", line 3258, in __getitem__
    value = next(
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/projects/sems/spack/lib/spack/spack/build_environment.py", line 847, in child_process
    return_value = function()
  File "/projects/sems/spack/lib/spack/spack/installer.py", line 1149, in build_process
    phase(pkg.spec, pkg.prefix)
  File "/projects/sems/spack/lib/spack/spack/package.py", line 109, in phase_wrapper
    phase(spec, prefix)
  File "/Users/mmundt/tpl_building/sems-spack-tpls/src/project/dev/dev-repo/packages/vvtest/package.py", line 28, in install
    python('install_vvtest', prefix.bin)
  File "/projects/sems/spack/lib/spack/spack/spec.py", line 3267, in __getitem__
    raise KeyError("No spec with name %s in %s" % (name, self))
KeyError: 'No spec with name python in vvtest@1.2.0%apple-clang@11.0.3 arch=darwin-catalina-x86_64'

==> [2020-09-17-10:48:56.956277] ChildError: KeyError: 'No spec with name python in vvtest@1.2.0%apple-clang@11.0.3 arch=darwin-catalina-x86_64'
/Users/mmundt/tpl_building/sems-spack-tpls/src/project/dev/dev-repo/packages/vvtest/package.py:27, in install:
         26    def install(self, spec, prefix):
  >>     27        python = spec['python'].command
         28        python('install_vvtest', prefix.bin)

See build log for details:
  /var/folders/zz/zyxvpxvq6csfxvn_n000_dw8002lg2/T/mmundt/spack-stage/spack-stage-vvtest-1.2.0-y56i7a72h5tfn3vq46i7y2i23qtwkig2/spack-build-out.txt
Traceback (most recent call last):
  File "/projects/sems/spack/lib/spack/spack/spec.py", line 3258, in __getitem__
    value = next(
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/projects/sems/spack/lib/spack/spack/build_environment.py", line 847, in child_process
    return_value = function()
  File "/projects/sems/spack/lib/spack/spack/installer.py", line 1149, in build_process
    phase(pkg.spec, pkg.prefix)
  File "/projects/sems/spack/lib/spack/spack/package.py", line 109, in phase_wrapper
    phase(spec, prefix)
  File "/Users/mmundt/tpl_building/sems-spack-tpls/src/project/dev/dev-repo/packages/vvtest/package.py", line 28, in install
    python('install_vvtest', prefix.bin)
  File "/projects/sems/spack/lib/spack/spack/spec.py", line 3267, in __getitem__
    raise KeyError("No spec with name %s in %s" % (name, self))
KeyError: 'No spec with name python in vvtest@1.2.0%apple-clang@11.0.3 arch=darwin-catalina-x86_64'

```

Reply:
Can you run `spack spec vvtest`?

Reply:
```
$ spack spec vvtest
Input spec
--------------------------------
vvtest

Concretized
--------------------------------
vvtest@1.2.0%apple-clang@11.0.3 arch=darwin-catalina-x86_64
    ^python@3.8.5%apple-clang@11.0.3+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87 arch=darwin-catalina-x86_64
        ^bzip2@1.0.8%apple-clang@11.0.3+shared arch=darwin-catalina-x86_64
            ^diffutils@3.7%apple-clang@11.0.3 arch=darwin-catalina-x86_64
                ^libiconv@1.16%apple-clang@11.0.3 arch=darwin-catalina-x86_64
        ^expat@2.2.9%apple-clang@11.0.3~libbsd arch=darwin-catalina-x86_64
        ^gdbm@1.18.1%apple-clang@11.0.3 arch=darwin-catalina-x86_64
            ^readline@8.0%apple-clang@11.0.3 arch=darwin-catalina-x86_64
                ^ncurses@6.2%apple-clang@11.0.3~symlinks+termlib arch=darwin-catalina-x86_64
                    ^pkgconf@1.7.3%apple-clang@11.0.3 arch=darwin-catalina-x86_64
        ^gettext@0.21%apple-clang@11.0.3+bzip2+curses+git~libunistring+libxml2+tar+xz arch=darwin-catalina-x86_64
            ^libxml2@2.9.10%apple-clang@11.0.3~python arch=darwin-catalina-x86_64
                ^xz@5.2.5%apple-clang@11.0.3~pic arch=darwin-catalina-x86_64
                ^zlib@1.2.11%apple-clang@11.0.3+optimize+pic+shared arch=darwin-catalina-x86_64
            ^tar@1.32%apple-clang@11.0.3 arch=darwin-catalina-x86_64
        ^libffi@3.3%apple-clang@11.0.3 patches=26f26c6f29a7ce9bf370ad3ab2610f99365b4bdd7b82e7c31df41a3370d685c0 arch=darwin-catalina-x86_64
        ^openssl@1.1.1g%apple-clang@11.0.3+systemcerts arch=darwin-catalina-x86_64
            ^perl@5.30.3%apple-clang@11.0.3+cpanm+shared+threads arch=darwin-catalina-x86_64
                ^berkeley-db@18.1.40%apple-clang@11.0.3 arch=darwin-catalina-x86_64
        ^sqlite@3.33.0%apple-clang@11.0.3+column_metadata+fts~functions~rtree arch=darwin-catalina-x86_64
```

Reply:
Hmm, so it depends on python but it can't find python. I have no idea what's going on, I've never seen anything like this before. Can you run `spack debug report`?

Reply:
Much like with doctors, it's the worst to be the first one to find something unusual:
```
$ spack debug report
* **Spack:** 0.15.4-948-e7040467f
* **Python:** 3.8.5
* **Platform:** darwin-catalina-ivybridge
```

Reply:
@scheibelp can you look at this? I'm wondering if this is an additional Python 3.8 + macOS bug that we didn't know about. Either that or I'm missing something simple here.